criu/mem.c

```diff
@@ -30,10 +30,257 @@
 #include "fault-injection.h"
 #include "prctl.h"
 #include "compel/infect-util.h"
+#include "compel/plugins/std/syscall-codes.h"
+#include "common/scm.h"
 
 #include "protobuf.h"
 #include "images/pagemap.pb-c.h"
 
+struct pidfd_entry {
+	pid_t				pid;
+	int				pidfd;
+	struct hlist_node		hash; /* To lookup pidfd by pid */
+};
+
+static int pidfd_store_sk = -1;
+#define PIDFD_HASH_SIZE	32
+static struct hlist_head pidfd_hash[PIDFD_HASH_SIZE];
+
+int init_pidfd_store_sk(pid_t pid, int sk)
+{
+	int pidfd;
+	int sock_type;
+	socklen_t len;
+	struct sockaddr_un addr;
+	unsigned int addrlen;
+	/* In kernel a bufsize has type int and a value is doubled. */
+	uint32_t buf[2] = { INT_MAX / 2, INT_MAX / 2 };
+
+	if (!kdat.has_pidfd_open) {
+		pr_err("pidfd_open syscall is not supported");
+		return -1;
+	}
+
+	if (!kdat.has_pidfd_getfd) {
+		pr_err("pidfd_getfd syscall is not supported");
+		return -1;
+	}
+
+	/* Steal pidfd store socket from RPC client */
+	pidfd = syscall(SYS_pidfd_open, pid, 0);
+	if (pidfd == -1) {
+		pr_perror("Can't get pidfd of (pid: %d)", pid);
+		goto err;
+	}
+
+	pidfd_store_sk = syscall(SYS_pidfd_getfd, pidfd, sk, 0);
+	close(pidfd); /* Not needed anymore */
```

Comment:
This close overrides the errno of syscall. And the pr_perror below would print wrong errno. Probably it's better to instead do:

```
+       pidfd_store_sk = syscall(SYS_pidfd_getfd, pidfd, sk, 0);
+       if (pidfd_store_sk == -1) {
+               pr_perror("Can't steal fd %d using pidfd_getfd", sk);
+               close(pidfd);
+               goto err;
+       }
+       close(pidfd);
```